src/databricks/labs/dqx/llm/llm_core.py

```diff
@@ -0,0 +1,304 @@
+        )
+
+        # Validate and clean the JSON output
+        if hasattr(result, 'quality_rules'):
+            try:
+                # Try to parse the JSON to ensure it's valid
+                parsed_rules = json.loads(result.quality_rules)
+                # Re-serialize to ensure consistent formatting
+                result.quality_rules = json.dumps(parsed_rules, separators=(',', ':'))
+            except json.JSONDecodeError as e:
+                logger.warning(f"Generated invalid JSON: {e}. Raw output: {result.quality_rules}")
+                # Return a fallback empty array if JSON is invalid
+                result.quality_rules = "[]"
+
+        return result
+
+    def get_json_rules(self, schema_info: str, business_description: str, available_functions: str) -> str:
+        """
+        Get rules as a JSON string.
+
+        Args:
+            schema_info (str): JSON string containing table schema with column names, types, and sample data.
+            business_description (str): Natural language description of data quality requirements.
+            available_functions (str): JSON string of available DQX check functions.
+
+        Returns:
+            str: A JSON string representing the generated data quality rules.
+        """
+        result = self.forward(schema_info, business_description, available_functions)
+        return result.quality_rules
+
+
+def _configure_dspy_model(model: str, api_key: str = "", api_base: str = ""):
+    """
+    Configure the Dspy language model.
+
+    Args:
+        model (str): The model to use for the Dspy language model.
+        api_key (str): The API key for the model. Not required by Databricks foundational models.
+        api_base (str): The API base URL for the model. Not required by Databricks foundational models.
+    """
+    language_model = dspy.LM(
+        model=model,
+        model_type="chat",
+        api_key=api_key,
+        api_base=api_base,
+        max_retries=3,
+    )
+    dspy.configure(lm=language_model)
+
+
+def validate_generated_rules(expected: str, actual: str) -> float:
+    """
+    Validate generated rules with granular scoring for better optimizer feedback.
+
+    Scoring breakdown:
+        - JSON parsing (20%): Checks if the actual output can be parsed as valid JSON.
+        - Rules validation (50%): Ensures the rules pass DQX validation checks.
+        - Content similarity (30%): Compares the actual rules content with the expected rules.
+
+    Args:
+        expected (str): JSON string of the expected rules.
+        actual (str): JSON string of the actual generated rules.
+
+    Returns:
+        float: A score between 0.0 and 1.0 representing the quality of the generated rules.
+    """
+    total_score = 0.0
+
+    # Score weights
+    json_weight = 0.2
+    rule_weight = 0.5
+    similarity_weight = 0.3
+
+    expected_rules = json.loads(expected)
+
+    # Json parsing score (20%)
+    try:
+        actual_rules = json.loads(actual)
```

Comment:
this does not seem to be necessary since the `forward` method in `DQRuleGeneration` is making sure a valid json is outputed